services/sky/document_view.cc

```diff
@@ -99,30 +95,64 @@ base::WeakPtr<DocumentView> DocumentView::GetWeakPtr() {
   return weak_factory_.GetWeakPtr();
 }
 
-void DocumentView::OnEmbed(
-    mojo::View* root,
-    mojo::InterfaceRequest<mojo::ServiceProvider> services_provided_to_embedder,
-    mojo::ServiceProviderPtr services_provided_by_embedder) {
-  root_ = root;
+void DocumentView::InitViewport() {
+  mojo::ServiceProviderPtr viewport_service_provider;
+  shell_->ConnectToApplication("mojo:native_viewport_service",
+                               mojo::GetProxy(&viewport_service_provider),
+                               nullptr);
+  mojo::ConnectToService(viewport_service_provider.get(), &viewport_service_);
+  viewport_service_.set_connection_error_handler(
+      [this]() { OnViewportConnectionError(); });
```

Comment:
The lifetime of viewport_service_ is bound to that of this object so it will not outlive it.